shared/subspace-metrics/src/lib.rs

```diff
@@ -82,11 +82,35 @@ pub fn start_prometheus_metrics_server(
     let shared_registry = Arc::new(Mutex::new(registry));
     let data = Data::new(shared_registry);
 
-    info!(?endpoints, "Starting metrics server...",);
+    let app_factory = move || App::new().app_data(data.clone()).service(metrics);
+    let result = HttpServer::new(app_factory.clone()).bind(endpoints.as_slice());
 
-    Ok(
-        HttpServer::new(move || App::new().app_data(data.clone()).service(metrics))
-            .bind(endpoints.as_slice())?
-            .run(),
-    )
+    let server = match result {
+        Ok(server) => server,
+        Err(error) => {
+            warn!(
+                ?error,
+                "Failed to start metrics server. Fallback to the random port on localhost.",
+            );
+
+            let default_endpoint = SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 0);
+            let result = HttpServer::new(app_factory).bind(default_endpoint);
```

Comment:
Why not simply `?` here?

Reply:
The default error message is rather cryptic.